File: packages/entangle-udp/ios/EntangleUdpModule.swift
import ExpoModulesCore
import Network

/// A datagram socket for pointer frames.
///
/// `send` is a synchronous `Function` rather than an `AsyncFunction` on
/// purpose: this is called up to 120 times a second, and a promise per frame
/// would cost more than the frame. `NWConnection.send` hands the bytes to its
/// own queue, so the call returns immediately.
public class EntangleUdpModule: Module {
  private var connection: NWConnection?
  private let queue = DispatchQueue(label: "entangle.udp.client", qos: .userInteractive)

  public func definition() -> ModuleDefinition {
    Name("EntangleUdp")

    OnDestroy {
      self.closeConnection()
    }

    Function("open") { (host: String, port: Int) -> Bool in
      self.closeConnection()
      guard port > 0, port <= 65535,
            let endpointPort = NWEndpoint.Port(rawValue: UInt16(port)) else {
        return false
      }
      let connection = NWConnection(
        host: NWEndpoint.Host(host), port: endpointPort, using: .udp
      )
      connection.start(queue: self.queue)
      self.connection = connection
      return true
    }

    Function("send") { (text: String) -> Bool in
      guard let connection = self.connection else { return false }
      // `.idempotent` skips the per-send completion block. A dropped frame is
      // the transport working as intended: the next one carries the total.
      connection.send(content: Data(text.utf8), completion: .idempotent)
      return true
    }

    Function("close") { () -> Void in
      self.closeConnection()
    }
  }

  private func closeConnection() {
    connection?.cancel()
    connection = nil
  }
}

File: packages/entangle-udp/android/src/main/java/expo/modules/entangleudp/EntangleUdpModule.kt
package expo.modules.entangleudp

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.util.concurrent.Executors

/**
 * A datagram socket for pointer frames.
 *
 * `send` is synchronous from JavaScript's point of view but does no I/O on the
 * calling thread: the packet goes to a single-threaded executor. This is
 * called up to 120 times a second and must never block the JS thread, which is
 * already the narrowest part of the pointer path.
 */
class EntangleUdpModule : Module() {
  private val executor = Executors.newSingleThreadExecutor()
  private var socket: DatagramSocket? = null
  @Volatile private var address: InetAddress? = null
  @Volatile private var port: Int = 0

  override fun definition() = ModuleDefinition {
    Name("EntangleUdp")

    OnDestroy {
      closeSocket()
    }

    Function("open") { host: String, targetPort: Int ->
      closeSocket()
      if (targetPort <= 0 || targetPort > 65535) {
        return@Function false
      }
      port = targetPort
      // Resolving can touch the network — a `.local` name goes to mDNS — so it
      // happens off the calling thread. `send` no-ops until it lands.
      executor.execute {
        try {
          address = InetAddress.getByName(host)
          socket = DatagramSocket()
        } catch (error: Exception) {
          address = null
          socket = null
        }
      }
      true
    }

    Function("send") { text: String ->
      val target = address ?: return@Function false
      val bytes = text.toByteArray(Charsets.UTF_8)
      val currentPort = port
      executor.execute {
        try {
          socket?.send(DatagramPacket(bytes, bytes.size, target, currentPort))
        } catch (error: Exception) {
          // A dropped frame is the transport working as intended: the next one
          // carries the running total.
        }
      }
      true
    }

    Function("close") {
      closeSocket()
    }
  }

  private fun closeSocket() {
    address = null
    port = 0
    val current = socket
    socket = null
    executor.execute {
      try {
        current?.close()
      } catch (error: Exception) {
      }
    }
  }
}
